edgehog-device-runtime-forwarder/src/connections_manager.rs

```diff
@@ -0,0 +1,281 @@
+    rx_ws: UnboundedReceiver<ProtoMessage>,
+    /// bridge URL.
+    url: Url,
+}
+
+impl ConnectionsManager {
+    /// Establish a new WebSocket connection between the device and the bridge.
+    #[instrument]
+    pub async fn connect(url: Url) -> Result<Self, Error> {
+        // TODO: check if, when a wrong URL is passed, it will endlessly try to connect
+
+        let ws_stream = Self::ws_connect(&url).await?;
+
+        // this channel is used by tasks associated to each connection to communicate new
+        // information available on a given websocket between the device and TTYD.
+        // it is also used to forward the incoming data from TTYD to the device.
+        let (tx_ws, rx_ws) = unbounded_channel();
+
+        let connections = Connections::new(tx_ws);
+
+        Ok(Self {
+            connections,
+            ws_stream,
+            rx_ws,
+            url,
+        })
+    }
+
+    /// Perform exponential backoff while trying to (re)connect with the bridge.
+    #[instrument(skip_all)]
+    pub(crate) async fn ws_connect(
+        url: &Url,
+    ) -> Result<WebSocketStream<MaybeTlsStream<TcpStream>>, Error> {
+        // try openning a websocket connection with the bridge using exponential backoff
+        let (ws_stream, http_res) =
+            backoff::future::retry(ExponentialBackoff::default(), || async {
+                debug!("creating websocket connection with {}", url);
+                Ok(connect_async(url).await?)
+            })
+            .await
+            .map_err(Error::WebSocket)?;
+
+        trace!("bridge websocket response {http_res:?}");
+
+        Ok(ws_stream)
+    }
+
+    /// Manage the reception and transmission of data between the websocket and each connection.
+    ///
+    /// It performs specific operations depending on the occurrence of one of the following events:
+    /// * Receiving data from the WebSocket,
+    /// * A timeout event occurring before any data is received from the WebSocket connection,
+    /// * Receiving data from one of the connections (e.g., between the device and TTYD).
+    #[instrument(skip_all)]
+    pub async fn handle_connections(&mut self) -> Result<(), Error> {
+        loop {
+            match self.event_loop().await {
+                Ok(ControlFlow::Continue(())) => {}
+                Ok(ControlFlow::Break(())) => break,
+                // if the connection has been suddenly interrupted, try re-establishing it.
+                // only Tungstenite errors should be handled for device reconnection
+                Err(err) => {
+                    error!("WebSocket error {err:?}");
+                    self.reconnect().await?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Handle a single connection event.
+    #[instrument(skip_all)]
+    pub(crate) async fn event_loop(&mut self) -> Result<ControlFlow<()>, TungError> {
+        let event = self
+            .select_ws_event(Duration::from_secs(PING_TIMEOUT))
```

Comment:
Make the timeout `Duration` a const since from_secs is const (since 1.32) and pass it directly to tokio::time::timeout (in the select) instead of constructing it each iteration and passing it to the function, since duration is Copy